### PR TITLE
Refactor data loading

### PR DESCRIPTION
**Pull request by kris-lab**

### CODE REVIEW

**Comment by kris-lab** [comment]
@tomaszdurka could you have a look? it is working and stable now.

File: lib/mms/cli.rb [hunk 1]
```diff
@@ -158,7 +155,9 @@ class MMS::CLI::Command::Clusters < MMS::CLI::Command
       subcommand 'list', 'Cluster list' do
 
         def execute
-          cluster_list = agent.clusters
+          group_id = ignore? ? nil : @config.default_group_id
+
+          cluster_list = agent.clusters(group_id)
```
Comment (tomaszdurka) [inline lib/mms/cli.rb]: I've noticed you've altered methods on agent to be more flexible and allow passing of group_id everywhere.
Wdyt about reverting that and converting above lines to:
```ruby
group_id = group_id || @config.default_group_id
if group_id and !ignore 
  cluster_list = agent.find_group(group_id).clusters
else
  cluster_list = agent.clusters
end
```
?

That would be applicable to other commands (hosts, clusters, groups).

Comment (tomaszdurka) [inline lib/mms/cli.rb]: Additionally can `group_id` be more complicated getter/method and already consider `@config.group_id` to not repeat that code in each command?

Comment (tomaszdurka) [inline lib/mms/cli.rb]: I can see config is set in initialize so maybe having method `group` on abstract command would simplify a lot?
```ruby
def group
  return nil if ignore
  group_id = group_id || @config.group_id
  return MMS::Resource::Group.new(group_id) if group_id
end
```

Comment (kris-lab) [inline lib/mms/cli.rb]: yep; basically agree but the problem is that:
- doing filtering at `agent` trigger less  http calls what makes all faster
- for snapshots/restorejobs where we loop over many data the limitation of groups is very handy. if we have it at `cli` then I would have to split some logic between `cli` and `agent``

not sure

Comment (tomaszdurka) [inline lib/mms/cli.rb]: Hmm, maybe there is small performance gain. I understand this is only one request less (no group to load). Additional thing is you lose validation of the group - if the group is actually invalid you will bascically get no clusters and no restorejobs (without errors). If group fails to load on start you will straight away that something is wront with it.

If it's only this one request for group and if/else in each cli command I vote for having it that way. But you decide.

Comment (kris-lab) [inline lib/mms/cli.rb]: I see your point and will try optimise. However we have to remember that it should be possible to create instance of e.g. `group` and call `save` which will really create resource on the API side and then we could create new `cluster`, `hosts` and so on on that `group`. I see use case on the `pulsar` site. MMS could be fully configured by pulsar task in the future.

**Comment by kris-lab** [comment]
@tomaszdurka what about implementing `*_list` on CLI so it can be adjusted to our usage. Agent will always get all possible resource?

**Comment by kris-lab** [comment]
@tomaszdurka so? can I merge that? I will open new issue with bumping to `0.1.0`.

File: lib/mms/cli.rb [hunk 2]
```diff
@@ -76,6 +76,35 @@ def agent
         @agent = MMS::Agent.new(client)
       end
 
+      def group_list
+        id = ignore? ? nil : @config.default_group_id
```
Comment (tomaszdurka) [inline lib/mms/cli.rb]: Are you taking into account option `group_id`?
What do you think about still implementing `group`?

File: lib/mms/cli.rb [hunk 3]
```diff
@@ -76,6 +76,35 @@ def agent
...
+        id.nil? ? agent.groups : [agent.findGroup(id)]
+      end
+
+      def cluster_list
+        id = ignore? ? nil : @config.default_cluster_id
```
Comment (tomaszdurka) [inline lib/mms/cli.rb]: Are you taking into account option `cluster_id`?
What do you think about still implementing `cluster`?

Comment (tomaszdurka) [inline lib/mms/cli.rb]: Although `cluster` method is a bit problematic as it needs to be within groups?
Can you use multiple values as `cluster_id` to limit selection to several clusters?

**Comment by tomaszdurka** [comment]
I'm not superhappy with all these list methods. What abou only keeping `groups` and `clusters` methods on cli? And implement rest on commands?

File: lib/mms/cli.rb (hunk 2 above)
Comment (kris-lab) [inline lib/mms/cli.rb]: - Which `group_id`? There is only `default-group-id` for general purpose and specific `group-id` for alert/restore create command.
- what group would do for us?

File: lib/mms/cli.rb (hunk 3 above)
Comment (kris-lab) [inline lib/mms/cli.rb]: - cluster is always part of group. cannot exist separately! what exactly do you mean?
- the limitation is ALL or ONE; there is no need to do it different. If you need such behaviour you can use gem `mms-api` and implement own application?

**Comment by kris-lab** [comment]
- moving implementation into command is ok with me; I would keep only group_list and cluster_list to avoid duplications.

File: lib/mms/cli.rb (hunk 2 above)
Comment (tomaszdurka) [inline lib/mms/cli.rb]: - Yes, I got confused - you set it on config not read later.
- Group would combine config.`default_group_id` with `group_id` I thought there is. But if there is none of these we don't need it.

Comment (tomaszdurka) [inline lib/mms/cli.rb]: I am such huge not-fan of tennary operators, but if you like them we can keep them. Would still see it more as:
```ruby
if ignore? or @config.default_group_id.nil?
  agent.groups
else
  agent.find_group(@config.default_group_id)
end
```

*Additional thing (rename `findGroup` to `find_group`)*

**Comment by kris-lab** [comment]
ok, I moved *_list into command itself. 

@tomaszdurka wdyt?

**Comment by tomaszdurka** [comment]
Looks good for me? And for you? :)

**Comment by kris-lab** [comment]
Well, I am happy if you are happy:) For me is ok, work very well with MMS restore from point in time and from snapshot.

In next step I test all one more time and then bump to 0.1.0.

**Pull request merged**